File: Kotlin/MatrixCalculation.kt
import kotlin.math.*

fun main() {
    var arr: MutableList<MutableList<Double>> = mutableListOf(mutableListOf(1.0,2.0), mutableListOf(3.0,4.0))
    MutableListPrint(Inverse(arr))
}

/*  MutableList Deep Copy Function from https://stackoverflow.com/questions/45206241/what-is-concise-way-of-deep-copying-a-2d-mutablelist  */
typealias NativeArray = java.lang.reflect.Array
@Suppress("UNCHECKED_CAST")
fun <T> T.deepCopy(): T {
    return when (this) {
        is Array<*> -> {
            val type = this.javaClass.componentType
            NativeArray.newInstance(type, size).also {
                this.forEachIndexed { i, item ->
                    NativeArray.set(it, i, item.deepCopy())
                }
            } as T
        }
        is MutableList<*> -> this.mapTo(mutableListOf()) { it.deepCopy() } as T
        is List<*> -> this.map { it.deepCopy() } as T
        is Cloneable -> this.javaClass.getDeclaredMethod("clone").let {
            it.isAccessible = true;
            it.invoke(this) as T
        }
        else -> this
    }
}
/*  MutableList Deep Copy Function  */

/*  Mutable List Visualize, edited from ArrayPrint function  */
fun MutableListPrint(arr: MutableList<*>, indent: Int = 0): String {
    var T = "    ".repeat(indent)
    var a = 0
    var str = ""
    for (i in arr) {
        a++
        if (i is MutableList<*>) {
            str = str + "$T#$a: \n" + MutableListPrint(i, indent+1)
        }
        else {
            str = str + "$T#$a: ${i}\n"
        }
    }
    if (indent == 0) println(str)
    return str
}
/*  Mutable List Visualize  */

fun ArrayProduct(n: MutableList<MutableList<Double>>, m: MutableList<MutableList<Double>>): MutableList<MutableList<Double>> {
    val res: MutableList<MutableList<Double>> = ArrayList()
    for (i in 0..n.size-1) {
        var a: MutableList<Double> = ArrayList()
        for (j in 0..m[0].size-1) {
            var q = 0.0
            for (k in 0..n[0].size-1) { q += n[i][k]*m[k][j] }
            a.add(q)
        }
        res.add(a)
    }
    return res
}

fun Inverse(n: MutableList<MutableList<Double>>): MutableList<MutableList<Double>> {
    val o = n.size
    fun Transpose(v: MutableList<MutableList<Double>>): MutableList<MutableList<Double>> {
        val res: MutableList<MutableList<Double>> = ArrayList()
        for (i in 0..o-1) {
            var a: MutableList<Double> = ArrayList()
            for (j in 0..o-1) { a.add(v[j][i]) }
            res.add(a)
        }
        
        return res
    }
    
    fun Cofactor(v: MutableList<MutableList<Double>>,I: Int,J: Int): MutableList<MutableList<Double>> {
        var q = v.deepCopy()
        for (i in q) { i.removeAt(J) }
        q.removeAt(I)
        
        return q
    }
    
    fun Abs(v: MutableList<MutableList<Double>>): Double {
        var w = v.size
        if (w == 1) { return v[0][0] }
        else if (w == 2) { return v[0][0]*v[1][1] - v[1][0]*v[0][1] }
        else {
            var res = 0.0
            for (i in 0..w-1) {
                res += v[0][i]*Abs(Cofactor(v,0,i))*((-1.0).pow(i))
            }
            return res
        }
        return 0.0
    }
    
    var D = 1.0/Abs(n)
    var Res: MutableList<MutableList<Double>> = ArrayList()
    for (i in 0..o-1) {
        var t: MutableList<Double> = ArrayList()
        for (j in 0..o-1) {
            t.add((-1.0).pow(i+j).toInt()*D*Abs(Cofactor(n,i,j)))
        }
        Res.add(t)
    }
    return Transpose(Res)
}

File: C#/MatrixCalculation.cs
using System;
using System.Collections.Generic;

class MainClass {
  public static void Main (string[] args) {
    List<double> a = new List<double> {1,2};
    List<double> b = new List<double> {3,4};
    Matrix m = new Matrix(new List<List<double>> {a,b});
    Console.WriteLine((m.pow(5)).text());
  }
}

class Matrix {
  public Matrix(List<List<double>> a) {
    m = a.Count;
    n = a[0].Count;
    arr = a;
  }

  public Matrix(int N) {
    m = N;
    n = N;
    arr = new List<List<double>>();
    for (int i=0; i<N; i++) {
      List<double> p = new List<double>();
      for (int j=0; j<N; j++) {
        if (i==j) { p.Add(1.0); }
        else { p.Add(0.0); }
      }
      arr.Add(p);
    }
  }

  public int m;
  public int n;
  public List<List<double>> arr;

  public static Matrix operator + (Matrix a, Matrix b) {
    int M = Math.Max(a.m, b.m);
    int N = Math.Max(a.n, b.n);
    List<List<double>> res = new List<List<double>>();

    for (int i=0; i<N; i++) {
      List<double> p = new List<double>();
      for (int j=0; j<M; j++) {
        p.Add(a.arr[i][j]+b.arr[i][j]);
      }
      res.Add(p);
    }

    return new Matrix(res);
  }

  public static Matrix operator * (double a, Matrix b) {
    List<List<double>> res = new List<List<double>>();

    for (int i=0; i<b.m; i++) {
      List<double> p = new List<double>();
      for (int j=0; j<b.n; j++) {
        p.Add(a*b.arr[i][j]);
      }
      res.Add(p);
    }

    return new Matrix(res);
  }

  public static Matrix operator * (Matrix a, double b) {
    return b*a;
  }

  public static Matrix operator * (Matrix a, Matrix b) {
    List<List<double>> res = new List<List<double>>();

    for (int i=0; i<a.m; i++) {
      List<double> p = new List<double>();
      for (int j=0; j<b.n; j++) {
        double Q = 0.0;
        for (int k=0; k<a.n; k++) {
          Q += a.arr[i][k]*b.arr[k][j];
        }
        p.Add(Q);
      }
      res.Add(p);
    }

    return new Matrix(res);
  }

  public Matrix pow(int n) {
    if (n==0) { return new Matrix(n); }
    if (n==1) { return new Matrix(arr); }
    return new Matrix(arr)*new Matrix(arr).pow(n-1);
  }

  public string text () {
    string res = "";
    string indent = "    ";
    for (int i=0; i<m; i++) {
      res += "#" + (i+1).ToString() + ": \n";
      for (int j=0; j<n; j++) {
        res += indent + "#" + (j+1).ToString() + ": " + arr[i][j].ToString() + "\n";
      }
    }

    return res;
  }
}
